osu.Game/Overlays/DirectOverlay.cs

```diff
@@ -232,21 +251,49 @@ protected override void PopIn()
         private SearchBeatmapSetsRequest getSetsRequest;
 
         private readonly Bindable<string> currentQuery = new Bindable<string>(string.Empty);
+        private int currentPage = 1;
+        private bool isLastPageLoaded = false;
 
         private ScheduledDelegate queryChangedDebounce;
+        private ScheduledDelegate addPageDebounce;
         private PreviewTrackManager previewTrackManager;
 
         private void queueUpdateSearch(bool queryTextChanged = false)
         {
             BeatmapSets = null;
             ResultAmounts = null;
+            currentPage = 1;
+            isLastPageLoaded = false;
 
             getSetsRequest?.Cancel();
 
             queryChangedDebounce?.Cancel();
             queryChangedDebounce = Scheduler.AddDelayed(updateSearch, queryTextChanged ? 500 : 100);
         }
 
+        private void queueAddPage()
+        {
+            if (getSetsRequest != null)
+                return;
+
+            if (addPageDebounce != null)
+                return;
+            
+            if (isLastPageLoaded)
+                return;
+            
+            currentPage++;
+            updateSearch();
+
+            getSetsRequest.Success += response =>
+            {
+                addPageDebounce = Scheduler.AddDelayed(() =>
+                {
+                    addPageDebounce = null;
+                }, 500);
```

Comment:
This feels dangerous/race-prone. I'm picturing a situation where the response is handled on a worker thread and the scheduling thread overwrites the outer assignment with the inner one. Or the other way around, really.

Reply:
I was thinking that this might be an issue, so I'm glad that you picked up on it. I'm thinking that I'll take your suggestion of making a separate paging component and try to handle this in there.

Reply:
Also, do you have any suggestions on handling this more elegantly? The main purpose of this whole debounce system is so that multiple page loads aren't triggered at once. In theory, what I really "want" to wait for is for the new `DirectPanel`s to be added so that `IsScrolledToEnd` is false. Am I best sticking with a more thread-safe debounce, or is there some other mechanism that I could use here?

Reply:
I mean I guess the debounce would be fine if it were guaranteed that both assignments are done on the scheduler thread. The obvious way to do this seems to me to be

```csharp
                Schedule(() => addPageDebounce = Scheduler.AddDelayed(() =>
                {
                    addPageDebounce = null;
                }, 500));
```

which however looks ugly (the innards might fare better as an extracted method) and also like it might catch fire due to nested schedule calls (although I *think* it should be fine, `SquirrelUpdateManager` also schedules a delayed task inside of a scheduled call).

FWIW I think the debounce call should stay within the overlay since that's presentation logic and belongs there.